packages/@sanity/form-builder/src/inputs/PortableText/text/Annotation.tsx

```diff
@@ -0,0 +1,308 @@
+  const {sanity} = useTheme()
+
+  // -------------- Popover ------------- //
+  const [textElement, setTextElement] = useState<HTMLSpanElement | null>(null)
+  const [open, setOpen] = useState<boolean>(false)
+  const [cursorRect, setCursorRect] = useState<DOMRect | null>(null)
+  const [selection, setSelection] = useState(null)
+  const isClosingRef = useRef<boolean>(false)
+  const rangeRef = useRef<Range | null>(null)
+
+  // This is a "virtual element" (supported by Popper.js)
+  const cursorElement = useMemo(() => {
+    if (!cursorRect) {
+      return null
+    }
+
+    return {
+      getBoundingClientRect: () => {
+        return cursorRect
+      },
+    }
+  }, [cursorRect])
+
+  useEffect(() => {
+    if (!open) {
+      return undefined
+    }
+
+    const handleScroll = () => {
+      if (rangeRef.current) {
+        setCursorRect(rangeRef.current.getBoundingClientRect())
+      }
+    }
+
+    scrollElement.addEventListener('scroll', handleScroll, {passive: true})
+
+    return () => {
+      scrollElement.removeEventListener('scroll', handleScroll)
+    }
+  }, [open, scrollElement])
+
+  // Close floating toolbar on Escape
+  useGlobalKeyDown(
+    useCallback(
+      (event) => {
+        if (event.key === 'Escape' && open) {
+          setOpen(false)
+        }
+      },
+      [open]
+    )
+  )
+
+  // Detect selection changes
+  useEffect(() => {
+    function handleSelectionChange() {
+      if (!textElement) return
+      const winSelection = window.getSelection()
+      const {anchorNode, anchorOffset, focusNode, focusOffset} = winSelection
+
+      setSelection({
+        anchorNode,
+        anchorOffset,
+        focusNode,
+        focusOffset,
+      })
+    }
+
+    document.addEventListener('selectionchange', handleSelectionChange, {passive: true})
+
+    return () => {
+      document.removeEventListener('selectionchange', handleSelectionChange)
+    }
+  }, [textElement])
+
+  // Open popover when selection is within annotations
+  useEffect(() => {
+    if (!selection) return
+    if (isClosingRef.current) return
+
+    const {anchorNode, focusNode} = selection
+
+    const annotationElement = annotationRef.current
+
+    if (annotationElement && annotationElement.contains(anchorNode) && anchorNode === focusNode) {
+      const range = window.getSelection().getRangeAt(0)
+      const rect = range.getBoundingClientRect()
+
+      rangeRef.current = range
+
+      if (rect) {
+        setCursorRect(rect)
+        setOpen(true)
+      }
+    } else {
+      setOpen(false)
+      setCursorRect(null)
+      rangeRef.current = null
+    }
+  }, [selection])
+  // -------------- End popover ------------- //
+
+  const text = useMemo(
+    () => (
+      <span ref={setTextElement} data-annotation="">
+        {children}
+      </span>
+    ),
+    [children]
+  )
+
+  const popoverScheme = useMemo(() => (sanity.color.dark ? 'light' : 'dark'), [sanity.color.dark])
```

Comment:
`useMemo` is overkill:
```suggestion
  const popoverScheme =sanity.color.dark ? 'light' : 'dark'
```